framework/calico_framework.py

```diff
@@ -0,0 +1,707 @@
+        for offer in offers:
+            _log.debug("RESERVING FOR FUTURE: %s, %s", offer, self)
+            if offer.reserved_by is None and offer.offer.slave_id.value in \
+                            [task.slave_id for task in self.tasks]:
+                offer.reserved_by = self
+
+    def timed_out(self, start=False):
+        """
+        Returns true if the test timed out.
+        :param start: Start the timer if it isn't currently running
+        """
+        if self.timeout is None:
+            if start:
+                self.restart_timeout()
+            return False
+        else:
+            return (time.time() - self.timeout) > TEST_TIMEOUT
+
+    def restart_timeout(self):
+        self.timeout = time.time()
+
+
+class Offer(object):
+    """
+    Provides encapsulation around the mesos offer for quick access to mesos
+    settings.
+    """
+
+    def __init__(self, offer):
+        self.offer = offer
+        self.reserved_by = None
+        self.tasks = []
+        self.cpus = 0.0
+        self.mem = 0.0
+        for resource in self.offer.resources:
+            if resource.name == "cpus":
+                self.cpus += resource.scalar.value
+            elif resource.name == "mem":
+                self.mem += resource.scalar.value
+            elif resource.name == "ports":
+                port_range = resource.ranges.range[:1].pop()
+                self.port = randint(port_range.begin, port_range.end)
+
+    def __eq__(self, other):
+        return self.offer.slave_id == other.offer.slave_id
+
+    def __hash__(self):
+        return hash(self.offer.slave_id.value)
+
+    @property
+    def offer_id(self):
+        return str(self.offer.id.value)
+
+    @property
+    def num_tasks_launchable(self):
+        return int(math.floor(min(self.cpus / TASK_CPUS, self.mem / TASK_MEM)))
+
+    def __repr__(self):
+        return "Offer(id=%s, tasks=%d, slave=%s, reserved_by=%s)" % (
+            self.offer_id[-5:], self.num_tasks_launchable,
+            self.offer.slave_id.value[-5:], self.reserved_by)
+
+
+class TestScheduler(mesos.interface.Scheduler):
+    def __init__(self, implicit_acknowledgements):
+        self.implicitAcknowledgements = implicit_acknowledgements
+        """
+        Flag to disable the requirement that the Executor responds to ACK
+        messages
+        """
+
+        self.framework_acks_sent = 0
+        """
+        Running total of how many acks have been sent to the executor
+        """
+
+        self.framework_acks_received = 0
+        """
+        Running total of how many acks have been received by the executor
+        """
+
+        self.tests = []
+        """
+        A collection of TestCases this scheduler will run through should deploy.
+        """
+
+        self.offers = set()
+        """
+        Keep track of offers available to this framework. Offers are unique by
+        their slave-id.
+        """
+
+    @property
```

Comment:
`@property` is a code smell.

It makes it seem to the caller that the operation is cheap, just dereferencing a name, when it could, in fact, be very expensive.  Just make it a method, it's purpose isn't to save you typing `()`!

The main time you should consider using it is when you refactor a class to have a different internal implementation, but need to keep the external API the same.